DB-11905/12121 Create a channel between StreamListener and ResultStreamer to pause a streaming if overloaded

ipraznik-splice opened the pull request:
<!--- Provide a general summary of your changes in the Title above. -->

## Short Description
The additional channel between StgreamListener and StreamableRDD to manage throttling 

## Long Description
The throttling mechanism we have currently, see StreamableRDD.submit(), is to run 2 concurrent jobs at a time on the spark side, each sending rows from 2 partitions to the region server. If the buffer for a particular partition is full, it would cause back-pressure, and prevent Spark from sending more rows for that partition. However, if the number of rows in the partition is smaller than the buffer size, then there is another mechanism to hold Spark from sending all partitions to the region server. 
There is additional channel from StreamableRDD to StreamListener. It uses the same hostname/port, which is passed with submit command and used to connect from ResultStreamer.
When the StreamListener has more than 3 * DEFAULT_PARALLEL_PARTITIONS in the queue, it sends the message PauseStream to StreamableRDD to pause to submit new partitions to Spark. If the consumption starts again, StreamListener sends the message ContinueStream to resume the partions submitting.

The functionality can be disabled by configuration parameter **spark.result.streaming.throttle.enabled**, by default is true.

## How to test
```
create table a ( i int);
insert into a values 1,2,3;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
```

```
get cursor c as 'select * from a --splice-properties useSpark=true
order by i';
next c;

ps -ef | grep 'com.splicemachine.test.SpliceTestPlatform' | less

jmap -histo:live [PID] | grep 'PartitionState'

```
It should not be bigger than 12 on the standalone cluster

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

cloudspliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2984/

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

cloudspliceci commented:
TEST SUCCEEDED +1

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

cloudspliceci commented:
TEST SUCCEEDED +1

hatyo commented on hbase_sql/src/main/java/com/splicemachine/stream/StreamListenerServer.java:
```diff
@@ -141,6 +141,14 @@ public void close() throws Exception {
             ctx.pipeline().remove(this);
+        } else if (msg instanceof StreamProtocol.InitOlapStream) {
+            StreamProtocol.InitOlapStream init = (StreamProtocol.InitOlapStream) msg;
+            LOG.trace("Received " + msg + " from " + channel);
```
ditto

ipraznik-splice replied:
done

dgomezferro commented on hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java:
```diff
@@ -403,6 +429,36 @@ public void failed(Throwable e) {
         return failure;
     }
+    
+    private void manageStreaming() {
```
Does this need synchronization? Is it possible to call `pauseStreaming` and `continueStreaming` almost at the same time and end up in a weird state? Like sending `StreamProtocol.PauseStream`, then `StreamProtocol.ContinueStream`, then setting `paused=false` and `pause=true` ?

ipraznik-splice replied:
added synchronization on pauseStreaming and continueStreaming

dgomezferro commented on hbase_sql/src/test/java/com/splicemachine/stream/StreamListenerIT.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.concurrent.*;
+
+@Category({LongerThanTwoMinutes.class, SlowTest.class, SerialTest.class})
+public class StreamListenerIT {
```
How long does this new IT take in jenkins?

dgomezferro replied:
It timed out after 280s in my laptop, is it possible to test this feature with a lighter weight test?

ipraznik-splice replied:
I've tested it locally, it failed only on 6th or 7th attempt. Takes about 2,5 mins.
On jenkins it's worse, failed on second attempt:

`2021-06-08|21:19:43,071 [INFO] Running com.splicemachine.stream.StreamListenerIT
2021-06-08|21:28:17,632 [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 514.474 s - in com.splicemachine.stream.StreamListenerIT

2021-06-09|19:35:09,450 [ERROR] testThrottlingNotBlocks(com.splicemachine.stream.StreamListenerIT)  Time elapsed: 246.612 s  <<< FAILURE!
java.lang.AssertionError: expected null, but was:<The statement has been cancelled or timed out.>
	at com.splicemachine.stream.StreamListenerIT.testThrottlingNotBlocks(StreamListenerIT.java:98)`

successful execution took about 5 mins

I've removed the test for now and will prepare better test in DB-12211

hatyo commented on hbase_sql/src/test/java/com/splicemachine/stream/StreamListenerIT.java:
```diff
@@ -0,0 +1,102 @@
+                            }
+                        }
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
```
I guess you can remove the `catch` since it is useless and leave `finally`.

ipraznik-splice replied:
This test is sporadic, I've removed it and will address it in another task

hatyo commented on hbase_sql/src/test/java/com/splicemachine/stream/StreamListenerIT.java:
```diff
@@ -0,0 +1,102 @@
+                        }
+                        rs.close();
+                        if (counter != TABLE_SIZE)
+                            return String.format("Counter does not matchL expected %d, received %d", TABLE_SIZE, counter);
```
typo: matchL

ipraznik-splice replied:
ditto

hatyo reviewed (commented):
Very clean and nice PR, great work @ipraznik-splice!
I reviewed your PR and have mostly minor comments.

dgomezferro reviewed (changes requested):
Overall it looks good, I only have some minor comments

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0

cloudspliceci commented:
TEST SUCCEEDED +1

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0

cloudspliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3094/

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0

cloudspliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3096/

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0

cloudspliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3101/

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0

cloudspliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3103/

ipraznik-splice commented on splice_access_api/src/main/java/com/splicemachine/access/configuration/OlapConfigurations.java:
```diff
@@ -190,8 +190,8 @@
     public static final String SPARK_SLOW_RESULT_STREAMING_BATCH_SIZE = "spark.slow.result.streaming.batch.size";
     public static final int DEFAULT_SPARK_SLOW_RESULT_STREAMING_BATCH_SIZE = 20;
 
-    public static final String SPARK_RESULT_STREAMING_THROTTLE_MAX_WAIT = "spark.result.streaming.throttle.max.wait";
-    public static final int DEFAULT_SPARK_RESULT_STREAMING_THROTTLE_MAX_WAIT = 120;
```
This parameter was introduced just a couple of weeks ago in the first version of this functionality to solve the issue with blocking executors, which is solved in the new implementatin. Don't think that somebody is using this parameter.

ipraznik-splice commented on hbase_sql/src/main/java/com/splicemachine/stream/OlapStreamListener.java:
```diff
@@ -0,0 +1,130 @@
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * OlapStreamListener handles a communication from StreamLister to manage throttling if a client overloaded with
```
fixed

ipraznik-splice commented on hbase_sql/src/main/java/com/splicemachine/stream/OlapStreamListener.java:
```diff
@@ -0,0 +1,130 @@
+                    clientConsuming = true;
+                    consumingCondition.signalAll();
+                }
+            }
```
added

ipraznik-splice commented on hbase_sql/src/main/java/com/splicemachine/stream/OlapStreamListener.java:
```diff
@@ -0,0 +1,130 @@
+            if (msg instanceof StreamProtocol.PauseStream) {
+                if (clientConsuming) {
+                    LOG.trace("Client is overloaded, asked for a pause: " + this);
+                    clientConsuming = false;
```
I've added a java doc for isClientConsuming and getCondition

ipraznik-splice commented on hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java:
```diff
@@ -112,10 +126,22 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+            //Main handler is in StreamListenerServer, but if InitOlapStream message
+            //comes later, than Init message and channel is redirected to the StreamListener
+            // already
+            LOG.trace("Received " + msg + " from " + channel);
```
done

ipraznik-splice commented on hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java:
```diff
@@ -112,10 +126,22 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
             partitionMap.remove(channel);
+        } else if (msg instanceof StreamProtocol.InitOlapStream) {
+            //Main handler is in StreamListenerServer, but if InitOlapStream message
+            //comes later, than Init message and channel is redirected to the StreamListener
```
done

ipraznik-splice commented:
> Very clean and nice PR, great work @ipraznik-splice!
> I reviewed your PR and have mostly minor comments.

thank you @hatyo, I've addressed your comments.

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

cloudspliceci commented:
TEST SUCCEEDED +1

carolp-503 reviewed (approved):
+1

martinrupp commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes
